lib/internal/crypto/aes.js

```diff
@@ -0,0 +1,335 @@
+  switch (name) {
+    case 'AES-CBC':
+      switch (length) {
+        case 128: return kKeyVariantAES_CBC_128;
+        case 192: return kKeyVariantAES_CBC_192;
+        case 256: return kKeyVariantAES_CBC_256;
+      }
+      break;
+    case 'AES-CTR':
+      switch (length) {
+        case 128: return kKeyVariantAES_CTR_128;
+        case 192: return kKeyVariantAES_CTR_192;
+        case 256: return kKeyVariantAES_CTR_256;
+      }
+      break;
+    case 'AES-GCM':
+      switch (length) {
+        case 128: return kKeyVariantAES_GCM_128;
+        case 192: return kKeyVariantAES_GCM_192;
+        case 256: return kKeyVariantAES_GCM_256;
+      }
+      break;
+    case 'AES-KW':
+      switch (length) {
+        case 128: return kKeyVariantAES_KW_128;
+        case 192: return kKeyVariantAES_KW_192;
+        case 256: return kKeyVariantAES_KW_256;
+      }
+      break;
+  }
+}
+
+function asyncAesCtrCipher(mode, key, data, { counter, length }) {
+  counter = getArrayBufferOrView(counter, 'algorithm.counter');
+  validateByteLength(counter, 'algorithm.counter', 16);
+  // The length must specify an integer between 1 and 128. While
+  // there is no default, this should typically be 64.
+  if (typeof length !== 'number' ||
+      length <= 0 ||
+      length > kMaxCounterLength) {
+    throw lazyDOMException(
+      'AES-CTR algorithm.length must be between 1 and 128',
+      'OperationError');
+  }
+
+  return jobPromise(new AESCipherJob(
+    kCryptoJobAsync,
+    mode,
+    key[kKeyObject][kHandle],
+    data,
+    getVariant('AES-CTR', key.algorithm.length),
+    counter,
+    length));
+}
+
+function asyncAesCbcCipher(mode, key, data, { iv }) {
+  iv = getArrayBufferOrView(iv, 'algorithm.iv');
+  validateByteLength(iv, 'algorithm.iv', 16);
+  return jobPromise(new AESCipherJob(
+    kCryptoJobAsync,
+    mode,
+    key[kKeyObject][kHandle],
+    data,
+    getVariant('AES-CBC', key.algorithm.length),
+    iv));
+}
+
+function asyncAesKwCipher(mode, key, data) {
+  return jobPromise(new AESCipherJob(
+    kCryptoJobAsync,
+    mode,
+    key[kKeyObject][kHandle],
+    data,
+    getVariant('AES-KW', key.algorithm.length)));
+}
+
+function asyncAesGcmCipher(
+  mode,
+  key,
+  data,
+  { iv, additionalData, tagLength = 128 }) {
+  if (!ArrayPrototypeIncludes(kTagLengths, tagLength)) {
+    throw lazyDOMException(
+      `${tagLength} is not a valid AES-GCM tag length`,
+      'OperationError');
+  }
+
+  iv = getArrayBufferOrView(iv, 'algorithm.iv');
+  validateMaxBufferLength(iv, 'algorithm.iv');
+
+  if (additionalData !== undefined) {
+    additionalData =
+      getArrayBufferOrView(additionalData, 'algorithm.additionalData');
+    validateMaxBufferLength(additionalData, 'algorithm.additionalData');
+  }
+
+  const tagByteLength = MathFloor(tagLength / 8);
+  let tag;
+  switch (mode) {
+    case kWebCryptoCipherDecrypt:
+      tag = data.slice(-tagByteLength);
+      data = data.slice(0, -tagByteLength);
```

Comment:
```suggestion
      tag = StringPrototypeSlice(data, -tagByteLength);
      data = StringPrototypeSlice(data, 0, -tagByteLength);
```

Reply:
These aren't strings, they are either ArrayBuffer or any TypedArray, none of which use interchangeable slice functions. Will work on it.

Reply:
The story with primordials and ArrayBuffer/DataView/TypedArray slice is a bit ... difficult at the moment. I'll investigate making this change after this lands as it's going to require some unrelated changes it appears.

Reply:
Sounds good!